packages/backend/src/Tabulators/IRV.ts

```diff
@@ -23,94 +42,190 @@ export function IRV(candidates: string[], votes: ballot[], nWinners = 1, randomT
+                remainingCandidates = [...summaryData.candidates].filter(c => !results.elected.includes(c))
+
+                // Reset candidate vote counts and redistribute votes
+                for (var i = 0; i < candidateVotes.length; i++) {
+                    candidateVotes[i] = [];
+                }
+                distributeVotes(remainingCandidates, candidateVotes, exhaustedVotes, weightedVotes)
+            }
+
+        }
+        else if ( proportional && remainingCandidates.length <= (nWinners - results.elected.length)) {
+            // If number of candidates remaining can fill seats, elect them and end election
+            // Only used in proportional as order can matter in block 
+            results.elected.push(...remainingCandidates)
+            roundResults.winners.push(...remainingCandidates)
         }
         else {
             // find candidate with least votes and remove from remaining candidates
-            let remainingVoteCounts = roundVoteCounts.filter((c, i) => remainingCandidatesIndexes.includes(i))
-            let lastPlaceCandidateIndex = remainingVoteCounts.indexOf(Math.min(...remainingVoteCounts))
-            remainingCandidates = remainingCandidates.filter(c => c !== summaryData.candidates[remainingCandidatesIndexes[lastPlaceCandidateIndex]])
+            let remainingVoteCounts = sortedVoteCounts.filter(c => remainingCandidatesIndexes.includes(c.index))
+            let lastPlaceCandidateIndex = remainingVoteCounts[remainingVoteCounts.length - 1].index
+            remainingCandidates = remainingCandidates.filter(c => c.index !== lastPlaceCandidateIndex)
+            let eliminatedCandidateVotes = candidateVotes[lastPlaceCandidateIndex]
+            candidateVotes[lastPlaceCandidateIndex] = []
+            distributeVotes(remainingCandidates, candidateVotes, exhaustedVotes, eliminatedCandidateVotes)
+            roundResults.eliminated.push(summaryData.candidates[lastPlaceCandidateIndex])
         }
+        results.roundResults.push(roundResults)
     }
-    // Sort data in order of candidate placements
-    // results.summaryData = sortData(summaryData, results.elected.concat(results.tied).concat(results.other))
+
     return results
 }
 
+function addWeightedVotes(weightedVotes: weightedVote[]) {
+    let voteTotal = new Fraction(0)
+    weightedVotes.forEach(vote => {
+        voteTotal = voteTotal.add(vote.weight)
+    })
+    return voteTotal
+}
+
+
+
+function distributeVotes(remainingCandidates: candidate[], candidateVotes: weightedVote[][], exhaustedVotes: weightedVote[], votesToDistribute: weightedVote[]) {
+    for (let i = votesToDistribute.length - 1; i >= 0; i--) {
+        // loop backwards over the ballots so exhausted votes can be removed
+        let ballot = votesToDistribute[i]
+        let topRemainingRank: number = 0
+        let topRemainingRankIndex: number = 0
+        let isOverVote = false
+        remainingCandidates.forEach(candidate => {
```

Comment:
nit: I think Array.reduce could be a good fit here

Reply:
Not really familiar with reduce outside of computing sums. I got something to work but it's not that clean given that it has to output both the rank and index. I'm going to leave it alone for now.